pytorch_lightning/trainer/trainer.py

```diff
@@ -500,20 +500,22 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
         """
         try:
             return trainer_fn(*args, **kwargs)
-        except KeyboardInterrupt:
+        except KeyboardInterrupt as exception:
             rank_zero_warn("Detected KeyboardInterrupt, attempting graceful shutdown...")
             # user could press Ctrl+c many times... only shutdown once
             if not self.interrupted:
                 self.state.status = TrainerStatus.INTERRUPTED
                 self.on_keyboard_interrupt()
-        except BaseException:
+                self.on_exception(exception)
```

Comment:
why here when it is already coved by the `KeyboardInterrupt`?

Reply:
good question, the idea is to deprecate on_keyboard_interrupt in the next PR https://github.com/daniellepintz/pytorch-lightning/pull/1 so this would cover the case where someone wants to add a hook for on_exception right now, which should work for all exceptions including KeyboardInterrupt

Reply:
if so, ten deprecate it in the same PR

Reply:
I initially had it in this PR, but @ananthsub recommended I move it out in https://github.com/PyTorchLightning/pytorch-lightning/pull/9183#pullrequestreview-741080873 . I can merge it back if that is preferred